### PR TITLE
[BEAM-2590] Implement basic Spark portable runner

### PR DESCRIPTION
**Pull request by ibzib**
This PR implements a prototype for a Spark portable runner.
- This work follows the Flink runner's example very closely.
- I chose to build this on top of the more stable and fully-featured legacy Spark runner rather than developing concurrently with the structured streaming branch. Therefore, RDDs are used rather than Datasets.
- This iteration implements and tests Impulse, Executable Stage, and GBK. Note that the GBK translation is mostly identical to the legacy runner.
- PAssert is not yet working (because metrics are not supported). However, `SparkPortableExecutionTest` is able to informally verify that the executable stage consumes and outputs elements correctly with a print statement.
- For now, streaming, side inputs, user state/timers, metrics, etc. are not supported. Side inputs are next on my road map.

R: @angoenka, @robertwb, @iemejia 

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: runners/spark/src/test/java/org/apache/beam/runners/spark/translation/SparkExecutableStageFunctionTest.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+import org.mockito.MockitoAnnotations;
+
+/** Unit tests for {@link SparkExecutableStageFunction}. */
+public class SparkExecutableStageFunctionTest {
+  @Mock private JobBundleFactoryCreator jobBundleFactoryCreator;
+  @Mock private JobBundleFactory jobBundleFactory;
+  @Mock private StageBundleFactory stageBundleFactory;
+  @Mock private RemoteBundle remoteBundle;
+
+  private final String INPUT_ID = "input-id";
```
Comment (iemejia) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/translation/SparkExecutableStageFunctionTest.java]: checkstyle not happy about this one.

Comment (ibzib) [reply runners/spark/src/test/java/org/apache/beam/runners/spark/translation/SparkExecutableStageFunctionTest.java]: Oops, fixed

**Comment by ibzib** [comment]
Run Python PreCommit

**Comment by iemejia** [comment]
R: Pinging other parties who could be interested in taking a look at this one @dmvk @amitsela

**Comment by iemejia** [comment]
Sorry @ibzib have not yet started the review of this one because I was finishing some stuff before the release cut, will take a look ASAP.

**Comment by ibzib** [comment]
Pinging @iemejia -- I would really like to hear your thoughts on this. I know it looks like a lot of code, but much of it is boilerplate. Most of the actual Spark logic is in [SparkBatchPortablePipelineTranslator.java](https://github.com/apache/beam/blob/897c1b2fe605d5a1338055e451e33608d490a65e/runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkBatchPortablePipelineTranslator.java), so that would be a good place to start looking.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkBatchPortablePipelineTranslator.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    Dataset inputDataset = context.popDataset(inputPCollectionId);
+    JavaRDD<WindowedValue<InputT>> inputRdd = ((BoundedDataset<InputT>) inputDataset).getRDD();
+    Map<String, String> outputs = transformNode.getTransform().getOutputsMap();
+    BiMap<String, Integer> outputMap = createOutputMap(outputs.values());
+
+    SparkExecutableStageFunction<InputT> function =
+        new SparkExecutableStageFunction<>(stagePayload, context.jobInfo, outputMap);
+    JavaRDD<RawUnionValue> staged = inputRdd.mapPartitions(function);
+
+    for (String outputId : outputs.values()) {
```
Comment (ibzib) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkBatchPortablePipelineTranslator.java]: I found there is an issue here--sometimes (e.g. for PAsserts) there will be no outputs, in which case we need to create a sink (like [this](https://github.com/ibzib/beam/blob/1f1660f0c97e65bb91e2c5aa7694b6d7b9215bbd/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchPortablePipelineTranslator.java#L398-L403)). I will add a fix in a future PR.

**Comment by iemejia** [comment]
Merged manually so closing.

**Pull request closed without merging**